test/unit/Cardano/DBLayer/MVarSpec.hs

```diff
@@ -1,14 +1,139 @@
+{-# LANGUAGE FlexibleInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
 module Cardano.DBLayer.MVarSpec
     ( spec
     ) where
 
 import Prelude
 
+import Cardano.DBLayer
+    ( DBLayer (..), PrimaryKey (..) )
 import Cardano.DBLayer.MVar
-    ()
+    ( newDBLayer )
+import Cardano.Wallet
+    ( Wallet (..), WalletId (..), initWallet )
+import Cardano.Wallet.Primitive
+    ( IsOurs (..) )
+import Control.Concurrent.Async
+    ( mapConcurrently_ )
+import Control.DeepSeq
+    ( NFData )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Data.List.NonEmpty
+    ( NonEmpty ((:|)) )
 import Test.Hspec
-    ( Spec )
+    ( Spec, describe, it, shouldBe )
+import Test.QuickCheck
+    ( Arbitrary (..), Property, checkCoverage, choose, vectorOf )
+import Test.QuickCheck.Monadic
+    ( monadicIO )
 
+import qualified Data.Set as Set
+import qualified Data.Text as T
 
 spec :: Spec
-spec = return ()
+spec = do
+    describe "DB works as expected" $ do
+        it "readCheckpoints works properly"
+            (checkCoverage dbReadCheckpointsProp)
+        it "replacement of values works properly"
+            (checkCoverage dbReplaceValsProp)
+        it "multiple sequential putCheckpoints work properly"
+            (checkCoverage dbMultiplePutsSeqProp)
+        it "multiple parallel putCheckpoints work properly"
+            (checkCoverage dbMultiplePutsParProp)
+
+
+{-------------------------------------------------------------------------------
+                                    Properties
+-------------------------------------------------------------------------------}
+
+
+dbReadCheckpointsProp
+    :: (PrimaryKey WalletId, Int)
+    -> Property
+dbReadCheckpointsProp (key, val)  = monadicIO $ liftIO $ do
+    db <- newDBLayer
+
+    putCheckpoints db key (toWalletState val)
+    resFromDb <- readCheckpoints db key
+
+    resFromDb `shouldBe` (Just $ toWalletState val)
+
+
+dbReplaceValsProp
+    :: (PrimaryKey WalletId, Int, Int)
+    -> Property
+dbReplaceValsProp (key, val1, val2)  = monadicIO $ liftIO $ do
+    db <- newDBLayer
+
+    putCheckpoints db key (toWalletState val1)
+    putCheckpoints db key (toWalletState val2)
+    resFromDb <- readCheckpoints db key
+
+    resFromDb `shouldBe` (Just $ toWalletState val2)
+
+dbMultiplePutsSeqProp
+    :: KeyValPairs
+    -> Property
+dbMultiplePutsSeqProp (KeyValPairs keyValPairs) = monadicIO $ liftIO $ do
+    db <- newDBLayer
+
+    mapM_ (\(key, val) -> putCheckpoints db key (toWalletState val)) keyValPairs
+    resFromDb <- Set.fromList <$> readWallets db
+
+    resFromDb `shouldBe` (Set.fromList (map fst keyValPairs))
+
+dbMultiplePutsParProp
+    :: KeyValPairs
+    -> Property
+dbMultiplePutsParProp (KeyValPairs keyValPairs) = monadicIO $ liftIO $ do
+    db <- newDBLayer
+
+    mapConcurrently_ (\(key, val) -> putCheckpoints db key (toWalletState val)) keyValPairs
+    resFromDb <- Set.fromList <$> readWallets db
```

Comment:
This makes me wonder if the DBLayer shouldn't just return a `Set` here in a first place :thinking:  ?

Reply:
for me using list suggests that the events/data are stored with "time order". For distributed systems with high throughput this is very high requirement, and usually it is better to start with `was stored` requirement (which implies `Set`). Probably, here at this point it does not matter. But to be honest, I do not know requirement here

Reply:
Hmmm.. That's a bit of a false assumption here. Especially because there's no particular ordering defined on the checkpoints and therefore, there's no guarantee that the DB layers will preserve the insertion order. 
Yet, one thing a `Set` will convey is the absence of duplicate in the data, which is right.